From a9f082b9e90cc769720631a7a783282d1e3a7f7f Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:06:06 -0400
Subject: [virt] kvm/ppc: book3s - Move skip-interrupt handlers to common code

Message-id: <1410545655-205645-138-git-send-email-dzickus@redhat.com>
Patchwork-id: 94673
O-Subject: [RHEL7 PATCH 137/626] KVM: PPC: Book3S: Move skip-interrupt handlers to common code
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 4f6c11db10159e362b0100d41b35bf6d731eb4e2
Author: Paul Mackerras <paulus@samba.org>
Date:   Fri Sep 20 14:52:50 2013 +1000

    KVM: PPC: Book3S: Move skip-interrupt handlers to common code

    Both PR and HV KVM have separate, identical copies of the
    kvmppc_skip_interrupt and kvmppc_skip_Hinterrupt handlers that are
    used for the situation where an interrupt happens when loading the
    instruction that caused an exit from the guest.  To eliminate this
    duplication and make it easier to compile in both PR and HV KVM,
    this moves this code to arch/powerpc/kernel/exceptions-64s.S along
    with other kernel interrupt handler code.

    Signed-off-by: Paul Mackerras <paulus@samba.org>
    Signed-off-by: Alexander Graf <agraf@suse.de>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kernel/exceptions-64s.S b/arch/powerpc/kernel/exceptions-64s.S
index 3a9ed6a..2a273be 100644
--- a/arch/powerpc/kernel/exceptions-64s.S
+++ b/arch/powerpc/kernel/exceptions-64s.S
@@ -650,6 +650,32 @@ slb_miss_user_pseries:
  b .    /* prevent spec. execution */
 #endif /* __DISABLED__ */
 
+#ifdef CONFIG_KVM_BOOK3S_64_HANDLER
+kvmppc_skip_interrupt:
+ /*
+  * Here all GPRs are unchanged from when the interrupt happened
+  * except for r13, which is saved in SPRG_SCRATCH0.
+  */
+ mfspr r13, SPRN_SRR0
+ addi r13, r13, 4
+ mtspr SPRN_SRR0, r13
+ GET_SCRATCH0(r13)
+ rfid
+ b .
+
+kvmppc_skip_Hinterrupt:
+ /*
+  * Here all GPRs are unchanged from when the interrupt happened
+  * except for r13, which is saved in SPRG_SCRATCH0.
+  */
+ mfspr r13, SPRN_HSRR0
+ addi r13, r13, 4
+ mtspr SPRN_HSRR0, r13
+ GET_SCRATCH0(r13)
+ hrfid
+ b .
+#endif
+
 /*
  * Code from here down to __end_handlers is invoked from the
  * exception prologs above.  Because the prologs assemble the
diff --git a/arch/powerpc/kvm/book3s_hv_rmhandlers.S b/arch/powerpc/kvm/book3s_hv_rmhandlers.S
index b764c36..615ea6d 100644
--- a/arch/powerpc/kvm/book3s_hv_rmhandlers.S
+++ b/arch/powerpc/kvm/book3s_hv_rmhandlers.S
@@ -33,30 +33,6 @@
 #error Need to fix lppaca and SLB shadow accesses in little endian mode
 #endif
 
-/*****************************************************************************
- *                                                                           *
- *        Real Mode handlers that need to be in the linear mapping           *
- *                                                                           *
- ****************************************************************************/
-
- .globl kvmppc_skip_interrupt
-kvmppc_skip_interrupt:
- mfspr r13,SPRN_SRR0
- addi r13,r13,4
- mtspr SPRN_SRR0,r13
- GET_SCRATCH0(r13)
- rfid
- b .
-
- .globl kvmppc_skip_Hinterrupt
-kvmppc_skip_Hinterrupt:
- mfspr r13,SPRN_HSRR0
- addi r13,r13,4
- mtspr SPRN_HSRR0,r13
- GET_SCRATCH0(r13)
- hrfid
- b .
-
 /*
  * Call kvmppc_hv_entry in real mode.
  * Must be called with interrupts hard-disabled.
diff --git a/arch/powerpc/kvm/book3s_rmhandlers.S b/arch/powerpc/kvm/book3s_rmhandlers.S
index cd59a3a..a38c4c9 100644
--- a/arch/powerpc/kvm/book3s_rmhandlers.S
+++ b/arch/powerpc/kvm/book3s_rmhandlers.S
@@ -38,32 +38,6 @@
 
 #define FUNC(name)   GLUE(.,name)
 
- .globl kvmppc_skip_interrupt
-kvmppc_skip_interrupt:
- /*
-  * Here all GPRs are unchanged from when the interrupt happened
-  * except for r13, which is saved in SPRG_SCRATCH0.
-  */
- mfspr r13, SPRN_SRR0
- addi r13, r13, 4
- mtspr SPRN_SRR0, r13
- GET_SCRATCH0(r13)
- rfid
- b .
-
- .globl kvmppc_skip_Hinterrupt
-kvmppc_skip_Hinterrupt:
- /*
-  * Here all GPRs are unchanged from when the interrupt happened
-  * except for r13, which is saved in SPRG_SCRATCH0.
-  */
- mfspr r13, SPRN_HSRR0
- addi r13, r13, 4
- mtspr SPRN_HSRR0, r13
- GET_SCRATCH0(r13)
- hrfid
- b .
-
 #elif defined(CONFIG_PPC_BOOK3S_32)
 
 #define FUNC(name)  name
-- 
1.7.1